Cesium3DTilesSelection/src/Tileset.cpp

```diff
@@ -2056,10 +2410,142 @@ void Tileset::processQueue(
+                TileContext* pContext = loadRecord.pTile->getContext();
+                if (pContext && pContext->implicitContext) {
+                  ImplicitTilingContext& implicitContext =
+                      *pContext->implicitContext;
+                  if (pQuadtreeID && implicitContext.quadtreeAvailability) {
+                    implicitContext.quadtreeAvailability->addLoadedSubtree(
+                        loadRecord.implicitInfo.pCurrentNode,
+                        std::move(*pSubtree.release()));
+                  } else if (pOctreeID && implicitContext.octreeAvailability) {
+                    implicitContext.octreeAvailability->addLoadedSubtree(
+                        loadRecord.implicitInfo.pCurrentNode,
+                        std::move(*pSubtree.release()));
+                  }
+                }
+              }
+            })
+        .catchInMainThread([this, id = std::move(id)](const std::exception& e) {
+          SPDLOG_LOGGER_ERROR(
+              this->_externals.pLogger,
+              "Unhandled error while loading the subtree for tile id {}: {}",
+              TileIdUtilities::createTileIdString(id),
+              e.what());
+          --this->_subtreeLoadsInProgress;
+        });
+  }
+}
+
+void Tileset::addSubtreeToLoadQueue(
+    Tile& tile,
+    ImplicitTraversalInfo&& implicitInfo,
+    double loadPriority) {
+
+  if (!implicitInfo.pCurrentNode &&
+      implicitInfo.availability & TileAvailabilityFlags::SUBTREE_AVAILABLE) {
+    TileContext* pContext = tile.getContext();
+    if (pContext && pContext->implicitContext) {
+      ImplicitTilingContext& implicitContext = *pContext->implicitContext;
+
+      const TileID& tileID = tile.getTileID();
+      const QuadtreeTileID* pQuadtreeID = std::get_if<QuadtreeTileID>(&tileID);
+      const OctreeTileID* pOctreeID = std::get_if<OctreeTileID>(&tileID);
+
+      if (pQuadtreeID && implicitContext.quadtreeAvailability) {
+        implicitInfo.pCurrentNode =
+            implicitContext.quadtreeAvailability->addNode(
+                *pQuadtreeID,
+                implicitInfo.pParentNode);
+      } else if (pOctreeID && implicitContext.octreeAvailability) {
+        implicitInfo.pCurrentNode = implicitContext.octreeAvailability->addNode(
+            *pOctreeID,
+            implicitInfo.pParentNode);
+      }
+
+      if (implicitInfo.pCurrentNode) {
+        this->_subtreeLoadQueue.push_back(
+            {&tile, std::move(implicitInfo), loadPriority});
+      }
+    }
+  }
+}
+
+void Tileset::processSubtreeQueue() {
+  if (this->_subtreeLoadsInProgress >=
+      this->_options.maximumSimultaneousSubtreeLoads) {
+    return;
+  }
+
+  std::sort(this->_subtreeLoadQueue.begin(), this->_subtreeLoadQueue.end());
+
+  for (SubtreeLoadRecord& record : this->_subtreeLoadQueue) {
+    // TODO: tracing code here
+    loadSubtree(std::move(record));
+    if (this->_subtreeLoadsInProgress >=
+        this->_options.maximumSimultaneousSubtreeLoads) {
+      break;
```

Comment:
Very possible I've missed something here, but if I'm reading this right... A subtree will only be added to `_subtreeLoadQueue` by `addSubtreeToLoadQueue` if the implicit info's `pCurrentNode` is null. And `addSubtreeToLoadQueue` _also_ sets `pCurrentNode` to no longer be null. So if we hit this break here, because we're loading too many subtrees, then the subtree will _never_ load. Right?

Reply:
Oof great catch here, `pCurrentNode` should be set to no longer null in `loadSubtree` instead, only once we're sure we're going to load it. This will mean that after `addSubtreeToLoadQueue`, `pCurrentNode` should still be null. Hopefully, that change works nicely with the rest of the implicit traversal state-space logic.